dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/thread/BackPressuredBlockingQueue.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.thread
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.core.configuration.BackPressureMitigation
+import com.datadog.android.core.configuration.BackPressureStrategy
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.TimeUnit
+
+internal class BackPressuredBlockingQueue<E : Any>(
+    private val logger: InternalLogger,
+    private val backPressureStrategy: BackPressureStrategy
+) : LinkedBlockingQueue<E>(
+    backPressureStrategy.capacity
+) {
+    override fun offer(e: E): Boolean {
+        return addWithBackPressure(e) {
+            @Suppress("UnsafeThirdPartyFunctionCall") // can't have NPE here
+            super.offer(it)
+        }
+    }
+
+    override fun offer(e: E, timeout: Long, unit: TimeUnit?): Boolean {
+        @Suppress("UnsafeThirdPartyFunctionCall") // can't have NPE here
+        val accepted = super.offer(e, timeout, unit)
+        if (!accepted) {
+            return offer(e)
+        } else {
+            if (remainingCapacity() == 0) {
+                onThresholdReached()
+            }
+            return true
+        }
+    }
+
+    private fun addWithBackPressure(
+        e: E,
+        operation: (E) -> Boolean
+    ): Boolean {
+        val remainingCapacity = remainingCapacity()
+        return if (remainingCapacity == 0) {
+            when (backPressureStrategy.backpressureMitigation) {
+                BackPressureMitigation.DROP_OLDEST -> {
+                    val first = take()
+                    onItemDropped(first)
+                    operation(e)
+                }
+
+                BackPressureMitigation.IGNORE_NEWEST -> {
+                    onItemDropped(e)
+                    true
+                }
+            }
+        } else {
+            if (remainingCapacity == 1) {
+                onThresholdReached()
```

Comment:
should we trigger this one? in theory, we may reach threshold, but there is no backpressure - by the time next element arrives it can fit into the queue.

Reply:
Yes, that's the point here to say that we reached the threshold, and that adding one more item _could_ lead to a loss of event. The idea is to have a callback before getting to that loss.